### ISSUE TITLE
[unitaryhack] Create a quantum simulator in PyTorch

### ISSUE THREAD

**Issue by co9olguy** [PennyLaneAI/pennylane#1225]
*This issue has been tagged for a $100 bounty during [unitaryHACK](https://unitaryfund.github.io/unitaryhack/participating-projects.html)*

PennyLane can currently perform quantum simulation (in a backprop-compatible way) using autograd, jax, and tensorflow. The notable exception is PyTorch. This has been hampered in the past by the fact that PyTorch did not support matrix multiplication or gradients for computaitons involving complex-valued tensors.

With the release of [PyTorch v1.8](https://pytorch.org/blog/pytorch-1.8-released/), it should now be possible to build a `default.qubit.torch` alongside `default.qubit.autograd`, `default.qubit.tf`, and `default.qubit.jax`. 

Thiis long-requested feature has been tackled before in #929, but that effort seems to have languished.

**Comment by anushkrishnav** [comment]
I would love to work on this.
Can you please share some guide or sample for me too look into and prep

**Comment by co9olguy** [comment]
Hi @anushkrishnav,
Awesome that you want to work on this :tada: 

Just a reminder that we've tagged it for [unitaryHACK](https://unitaryfund.github.io/unitaryhack/), which is running May 14-30th, so if you time your solution to come in during that period, and your PR is accepted, you can snag a bounty (and some street cred :sunglasses:)

If you want to prepare, the best places to start are:
- take a look at the [equivalent simulator device for TensorFlow](https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/devices/default_qubit_tf.py) to see how it works, as well as the [parent `default.qubit` device](https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/devices/default_qubit.py) (which already contains most functionality that will be needed
- check out the [release notes](https://github.com/pytorch/pytorch/releases) and related PRs to understand how PyTorch (as of 1.8) now supports complex numbers and complex-valued gradient computations

**Comment by Slimane33** [comment]
Hello, I am also interested to work on that. I was just wondering what was the general interest of building other `default.qubit` objects based on `tf`, `jax` or `torch` ? Is it to enable GPU running ?

**Comment by josh146** [comment]
Hi @Slimane33, yes! In particular, having a device written using TensorFlow/JAX/Torch has several advantages, since the entire computation is end-to-end written natively using the aforementioned framework. 

As a result,

- The entire computation is end-to-end differentiable using native backpropagation,
- The computation can take advantage of specific runtime features of the framework, such as GPU/TPU support, XLA acceleration, etc.

**Comment by antalszava** [comment]
With #1360 merged, closing this.
